Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/Meter/MeterMaintenance.tsx

```diff
@@ -0,0 +1,257 @@
+interface MaintenanceWindow {
+    ID: number,
+    MeterID: number,
+    StartTime: string,
+    EndTime: string
+}
+
+const MeterMaintenanceWindow = (props: IProps) => {
+    // Table Consts
+    const [sortKey, setSortKey] = React.useState<keyof MaintenanceWindow>('StartTime');
+    const [ascending, setAscending] = React.useState<boolean>(true);
+    const [status, setStatus] = React.useState<Application.Types.Status>('unintiated');
+    const [tableData, setTableData] = React.useState<MaintenanceWindow[]>([]);
+
+    // Modal Consts
+    const [showEditNew, setShowEditNew] = React.useState<boolean>(false);
+    const [showDeleteWarning, setShowDeleteWarning] = React.useState<boolean>(false);
+    const [activeWindow, setActiveWindow] = React.useState<MaintenanceWindow>(null);
+    const [reset, setReset] = React.useState<number>(0);
+
+    React.useEffect(() => {
+        let handle = getMaintenanceWindows();
+        return () => {
+            if (handle != null && handle.abort != null)
+                handle.abort();
+        };
+    }, [ascending, sortKey, props.Meter, reset]);
+
+    function getMaintenanceWindows(): JQuery.jqXHR<MaintenanceWindow[]> {
+        setStatus('loading');
+        let handle = $.ajax({
+            type: 'POST',
+            url: `${homePath}api/OpenXDA/MaintenanceWindow/SearchableList`,
+            contentType: "application/json; charset=utf-8",
+            data: JSON.stringify({
+                Searches: [{ FieldName: 'MeterID', Operator: "=", SearchText: props.Meter.ID, Type: 'string', isPivotColumn: false }],
+                OrderBy: sortKey,
+                Ascending: ascending
+            }),
+            dataType: 'json',
+            cache: false,
+            async: true
+        });
+        handle.done((data) => {
+            setTableData(JSON.parse(data.toString()));
+            setStatus('idle');
+        }).fail(() => setStatus('error'));
+        return handle;
+    }
+
+    function dbMaintenanceWindow(verb: 'POST' | 'DELETE' | 'PATCH', record: MaintenanceWindow): JQuery.jqXHR<MaintenanceWindow[]> {
+        setStatus('loading');
+        let action = (verb === 'POST') ? 'Add' : ((verb === 'DELETE') ? 'Delete' : 'Update');
+        let handle = $.ajax({
+            type: verb,
+            url: `${homePath}api/OpenXDA/MaintenanceWindow/${action}`,
+            contentType: "application/json; charset=utf-8",
+            data: JSON.stringify({ ...record }),
+            dataType: 'json',
+            cache: false,
+            async: true
+        });
+        handle.done((data) => {
+            setReset(reset + 1);
+        }).fail(() => setStatus('error'));
+        return handle;
+    }
+
+    interface ITimeProps<T> {
+        Record: T,
+        Setter: (T) => void,
+        Field: keyof T,
+        CheckLabel: string,
+        DateLabel: string,
+    }
+    function EnableTimeElement<T>(props: ITimeProps<T>): JSX.Element {
```

Comment:
why is this a `JSX.Element`? wouldn't `React.Fc<Iprops>` be more appropriate?
Also I'm not a big fan of a React Component being defined inside another Component. is there a reason for that nesting?

Reply:
React.FC doesn't allow generic props as far as I know. I could define this myself, but I'd like to keep functions as generic as possible in the case where this pattern crops up multiple times and we want to move this to gemstone.